### ISSUE TITLE
quantile random forest prediction for panel data

### ISSUE THREAD

**Issue by zpunit331111111** [grf-labs/grf#771]
Hello. I am using quantile_forest() to conduct prediction for a set of panel data, and get a bit confused. Here are my questions:
1. The individual IDs include characters, which raised an error when I use them for the 'clusters' parameter. Should I use only numbers, like 1 to 1000, to re-label them?  
2. I'd like to use the results from quantile_forest() to predict using a set of new data, should I include the ID as one independent variable or how should I tell the model which tree to use (assuming quantile_forest generate separated trees for each ID, right?)? I read the question #449, the answer said by passing cluster ID during training, trees will be associated cluster IDs, I am not sure how to match cluster ID from new data to the right trees. 
3. How the approach of using 'cluster' parameter to build one model different from training each cluster separately to build many models? I tried both ways, and they are different, but I can't figure out why. 
4. Under the same ID, I found the predicted values varied a lot. Since they all predict the same percentile (say, 80th), should I just average them as the final predicted 80th value? 
5.  Can "Cluster-Robust Estimation" predict using data that not belong to any clusters used for training?  In this paper: https://arxiv.org/pdf/1902.07409.pdf, section 1.2 indicated this can be done. However, the prediction results from clusters that have not been used for training were returned with NA. 

thank you so much!

**Comment by zpunit331111111** [comment]
Also a follow up question. Under the same ID, I found the predicted values varied a lot. Since they all predict the same percentile (say, 80th), should I just average them as the final predicted 80th value? thanks!

**Comment by erikcs** [comment]
Hi @zpunit331111111, 
> 1. The individual IDs include characters, which raised an error when I use them for the 'clusters' parameter. Should I use only numbers, like 1 to 1000, to re-label them?

Yes
> 2. I'd like to use the results from quantile_forest() to predict using a set of new data, should I include the ID as one independent variable or how should I tell the model which tree to use (assuming quantile_forest generate separated trees for each ID, right?)? I read the question #449, the answer said by passing cluster ID during training, trees will be associated cluster IDs, I am not sure how to match cluster ID from new data to the right trees.

Prediction examples are in the docstring: https://grf-labs.github.io/grf/reference/predict.quantile_forest.html
> 3. How the approach of using 'cluster' parameter to build one model different from training each cluster separately to build many models? I tried both ways, and they are different, but I can't figure out why.

Don't see why you expect this to be the same
> 4. Under the same ID, I found the predicted values varied a lot. Since they all predict the same percentile (say, 80th), should I just average them as the final predicted 80th value?

The predictions are conditional on X, if you want E[q80] instead of E[q80 | X=x] then you can average
> 5. Can "Cluster-Robust Estimation" predict using data that not belong to any clusters used for training?  In this paper: https://arxiv.org/pdf/1902.07409.pdf, section 1.2 indicated this can be done. However, the prediction results from clusters that have not been used for training were returned with NA.

Don't see why you want to do Cluster-Robust Estimation when your data does not belong to any cluster?

**Comment by zpunit331111111** [comment]
Hi thanks for your reply. For question 3, my understanding is that trees are built for each cluster separately, so I thought the results should be the same? That’s why I asked how the parameter ‘cluster’ is used in the algorithm. Is there any reference I can look up? Thanks!

**Comment by erikcs** [comment]
> Hi thanks for your reply. For question 3, my understanding is that trees are built for each cluster separately, so I thought the results should be the same? That’s why I asked how the parameter ‘cluster’ is used in the algorithm. Is there any reference I can look up? Thanks!
https://grf-labs.github.io/grf/REFERENCE.html#cluster-robust-estimation and last paragraph of 1.2 in https://arxiv.org/pdf/1902.07409.pdf

**Comment by zpunit331111111** [comment]
Thanks! For question 2, I don’t think the example actually considered cluster. As the example said, the new data should have the exact same variables as the training data, I still don’t know where to tell the model which cluster each data point in the new data belong to.

**Comment by erikcs** [comment]
Hi @zpunit331111111, there is no need to pass cluster membership for predictions, that information is only used under forest construction to identify the units of observation. From [the cluster reference above](https://grf-labs.github.io/grf/REFERENCE.html#cluster-robust-estimation): 

>... all subsequent calls to predict will take clusters into account, including when estimating the variance of forest predictions.

**Comment by zpunit331111111** [comment]
In question 5, I asked because I am wondering if the results could be generalized. For example, as said in the paper https://arxiv.org/pdf/1902.07409.pdf (last sentence on page 3). If I only used 76 schools to build the forest, and want to predict a new student for a new school. Since the new school doesn’t belong to any clusters, how will the prediction work?

**Comment by erikcs** [comment]
Prediction is not tied to clusters - forest subsampling and the unit of observation is. Closing as resolved, if anything is unclear please feel free to reopen.
